python/kachaka_api/aio/__init__.py

```diff
@@ -13,541 +13,234 @@
-    ) -> None:
-        self._robot_serial_number_callback = callback
-        if callback is not None:
-            asyncio.create_task(self._run_get_robot_serial_number())
 
-    def set_robot_version_callback(
-        self, callback: Optional[callable[[str], Awaitable[None]]]
-    ) -> None:
-        self._robot_version_callback = callback
-        if callback is not None:
-            asyncio.create_task(self._run_get_robot_version())
+T = TypeVar("T", bound=HasMetadata)
+U = TypeVar("U")
+P = ParamSpec("P")
+CallbackType = Callable[[U], Awaitable[None]] | None
+
 
-    def set_robot_pose_callback(
-        self, callback: Optional[callable[[Pose], Awaitable[None]]]
+class ResponseHandler(Generic[T, U]):
+    def __init__(
+        self,
+        get_function: Callable[[pb2.GetRequest], Awaitable[T]],
+        pick_response: Callable[[T], U],
     ) -> None:
-        self._robot_pose_callback = callback
+        self.callback: CallbackType[U] = None
+        self.get_function = get_function
+        self.pick_response = pick_response
+
+    async def _run(self) -> None:
+        cursor = 0
+        while self.callback is not None:
+            request = build_get_request(cursor)
+            response = await self.get_function(request)
+            cursor = response.metadata.cursor
+            if self.callback is not None:
+                picked_response = self.pick_response(response)
+                await self.callback(picked_response)
+
+    def set_callback(self, callback: CallbackType[U]) -> None:
+        if self.callback is not None and callback is not None:
+            raise RuntimeError("Callback is already set")
+        self.callback = callback
         if callback is not None:
-            asyncio.create_task(self._run_get_robot_pose())
+            asyncio.create_task(self._run())
+
 
-    def set_png_map_callback(
-        self, callback: Optional[callable[[Map], Awaitable[None]]]
+class TupleResponseHandler(ResponseHandler[T, U], Generic[T, U, P]):
+    def set_tuple_callback(
+        self, callback: Callable[P, Awaitable[None]] | None
     ) -> None:
-        self._png_map_callback = callback
-        if callback is not None:
-            asyncio.create_task(self._run_get_png_map())
+        if callback is None:
+            self.callback = None
+            return
+        if self.callback is not None:
+            raise RuntimeError("Callback is already set")
+
+        def extract_tuple_and_run(args: U) -> Awaitable[None]:
+            assert isinstance(args, tuple)
+            return callback(*args)
+
+        self.callback = extract_tuple_and_run
+        asyncio.create_task(self._run())
+
+
+class KachakaApiClient(KachakaApiClientBase):
+    def __init__(self, target: str = "100.94.1.1:26400") -> None:
+        super().__init__(target)
+        self._running = True
+
+        self.png_map: ResponseHandler[
+            pb2.GetPngMapResponse, pb2.Map
+        ] = ResponseHandler(self.stub.GetPngMap, lambda r: r.map)
```

Comment:
type hint を書くのではなくて、こう書くことはできませんか？
```suggestion
        self.png_map = ResponseHandler[
            pb2.GetPngMapResponse, pb2.Map
        ](self.stub.GetPngMap, lambda r: r.map)
```

Reply:
たしかに…！